charmcraft/declarative_plugin.py

```diff
@@ -0,0 +1,133 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2022 Canonical Ltd.
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License version 3 as published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Charmcraft's declarative plugin for craft-parts."""
+
+import io
+import pathlib
+import zipfile
+
+import jinja2
+import json
+import jsonschema
+import requests
+import yaml
+from craft_cli import emit, CraftError
+
+
+# TODO: probably this copyright should be removed/adapted
+METADATA_TEMPLATE = """# Copyright 2021 Ubuntu
+# See LICENSE file for licensing details.
+name: {{name}}
+description: |
+  {{description}}
+summary: |
+  {{summary}}
+containers:
+  application:
+    resource: application-image
+
+requires:
+{%- for relation in requires %}
+  {{ relation.name }}:
+    interface: {{ relation.interface }}
+{%- endfor %}
+
+resources:
+  application-image:
+    type: oci-image
+    description: |
+      OCI image containing the application to run.
+      Must have been built with Cloud Native Buildpacks (https://buildpacks.io)
+
+"""
+
+
+def prepare_charm(project_dir):
+    """Build a declarative charm.
+
+    Tthe build process is as follows:
+
+    - Get the remote base declarative charm with proper schema.
+
+    - Load the manifest and validate it with the retrieved schema.
+
+    - Generate the config files from manifest content.
+
+    - Pack the charm as a regular one.
+    """
+    emit.trace(f"Declarative building started from dir {project_dir}.")
+
+    manifest_filepath = project_dir / "manifest.yaml"
+    if not manifest_filepath.exists():
+        raise CraftError("Cannot find mandatory 'manifest.yaml' file.")
+    manifest_content = yaml.safe_load(manifest_filepath.read_text())
+    emit.trace("Manifest loaded ok.")
+
+    # TODO: we need to define a proper versioned location for this
```

Comment:
Big important decision here, about where the "cnb operator" will be located. Versions are important.